Feature/rvi Robust Variational Inference

Dashadower opened the pull request:
Check progress

Commit:
add structures for robust vi

Dashadower committed:
Merge branch 'develop' of https://github.com/stan-dev/stan into feature/rvi

Dashadower committed:
[WIP]Update gdpfit

Dashadower committed:
Merge branch 'develop' of https://github.com/stan-dev/stan into feature/rvi

Dashadower committed:
Implement gpdfit and update math submodule

Dashadower committed:
Add ess, rhat, mcse to advi

Dashadower committed:
Explicitly cast size() to size_t

Dashadower committed:
Merge branch 'develop' of https://github.com/stan-dev/stan into feature/rvi

Commit:
compute ess and mcse of from multiple chains

Dashadower committed:
Add function to returnvariational parameters, update run_rb. Work in progress.

Dashadower committed:
Merge branch 'develop' of https://github.com/stan-dev/stan into feature/rvi

Dashadower commented on src/stan/variational/advi.hpp:
```diff
@@ -522,7 +526,332 @@ class advi {
+   * @return Calculated MCSE
+   */
+
+  inline double ESS_MCSE(double &ess, double &mcse,
```
@hyunjimoon I'm sure the ess code is same as `compute_effective_sample_size()` in `src/stan/analyze/mcmc/compute_effective_sample_size.hpp`. Is there a reason why we can't use that function?

bbbales2 replied:
Definitely use that function if it does what you want

hyunjimoon replied:
To compute the variance of samples combined from all chains, `chain_sq_mean` was needed so I added the following to the original code. There might be a better way I guess.
```
Eigen::VectorXd chain_sq_mean(num_chains);
chain_sq_mean(chain) = draw.array().square().mean();
*mcse = std::sqrt((chain_sq_mean.mean() - chain_mean.mean() * chain_mean.mean())/ess_val);
```

Dashadower commented on src/stan/variational/base_family.hpp:
```diff
@@ -19,6 +19,16 @@ class base_family {
+  /**
+   * Return the approximation family's parameters as a single vector
+   */
+  virtual Eigen::VectorXd return_approx_params() const = 0;
```
Since meanfield and fullrank have different variational parameters, I though the best way to work with them is to just have the variational class return a single vector of variational parameters.

bbbales2 replied:
Makes sense for sure

Dashadower commented on src/stan/variational/families/normal_fullrank.hpp:
```diff
@@ -135,6 +135,37 @@ class normal_fullrank : public base_family {
+  /**
+   * Return the approximation family's parameters as a single vector
+   */
+  Eigen::VectorXd return_approx_params() const {
```
This part assumes that the only relevant parameters would be the lower triangular elements of the matrix. Am I correct?

bbbales2 replied:
Looks like it. I wouldn't worry so much about fullrank till the diagonal one is working tho.

Dashadower committed:
Remove redundant parentheses

Dashadower committed:
Fix typo

Dashadower committed:
Update rvi to store draws in row-major

Dashadower committed:
Update rvi

bbbales2 commented on src/stan/variational/advi.hpp:
```diff
@@ -522,7 +526,367 @@ class advi {
+  * sorting step.
+
+  */
+  static void gpdfit(Eigen::VectorXd& x, double& k, double& sigma, 
```
It looks like this function doesn't depend on any model stuff, so I would put it outside the model class somewhere. The same folder as the ESS/Rhat stuff would be good

bbbales2 commented on src/test/unit/variational/rvi_subroutine_test.cpp:
```diff
@@ -0,0 +1,113 @@
+}
+
+//TEST(RVI_SUBMODULE_TEST, gpdfit){
+//    // check values are equal to loo's gpdfit()
```
Comparing against loo seems like a good idea to me. Any reason these are commented out?

Dashadower replied:
Oh yeah that test is definitely included. I just ran it once or twice to make sure it passes in another file and forgot to uncomment it when copying it over.

bbbales2 committed:
Added more rvi tests

bbbales2 commented:
I added a couple tests and test models to go with them.

Start with this one:
```
./runTests.py test/unit/variational/rvi_std_normal_test
```

You're fitting a `normal(0, 1)` there and the idea is that you can slowly un-comment and add print statements to `run_rb` and make sure everything is working the way you expect.

There is a slightly harder test as well:

```
./runTests.py test/unit/variational/rvi_two_normals_test
```

On the algorithm side, should we be using RMSprop? From the last e-mail Jonathan sent "we use RMSProp as our base stochastic optimizer, so we can pretty much safely choose a step size 0.1 to start for any problem."

I was looking at the paper and it kinda hand-waved that the technique technically only applies to SGD but would probably work for these adaptive timestep things as long as the adaptation isn't based on things long in the past. If nobody knows we can ask.

On the C++ side, I've found it really handy to keep the clang sanitizer on lately (https://clang.llvm.org/docs/AddressSanitizer.html). It does a good job at catching memory errors. This is what my `make/local` looks like:

```
CXX=clang++
O=1
CXXFLAGS=-g -fsanitize=address -ftemplate-backtrace-limit=0 -fno-omit-frame-pointer
LDFLAGS=-g -fsanitize=address -fno-omit-frame-pointer
```

So if I get a segfault, there's already debugging enabled so I can run the debugger, and the sanitizer catches stuff that might not even totally blow up the program.

Dashadower commented:
@bbbales2 Thanks! They'll definitely be helpful. 
I believe what you said about rmsprop is true. Before we detect stationarity(so before SASA+ kicks in), we'd set the learning rate with RMSprop. I think Jonathan is saying that a-naught in SASA+ would be set to 0.1

Using an address sanitizer would be great since I'm working with alot of pointer arithmetic and it could hopefully catch where I messed up.

Dashadower committed:
Merge branch 'develop' of https://github.com/stan-dev/stan into feature/rvi

bbbales2 commented:
Yeah maybe the RMSprop starting with 0.1 is for once there is a SASA+-like component in there. I saw the `adapt_eta` function in there and that would definitely be a nice function to replace/clean up eventually. Leave it for now if it gets a good starting timestep tho.

Dashadower committed:
Update run_rvi, move gpdfit, create virtual set_approx_params in base_family.

Dashadower committed:
run_rvi update

Dashadower committed:
Update two normals test

Dashadower commented:
Okay. RVI will now run successfully when n_chains > 1 (I haven't finished up with split rhat for the single chain version to work).
I've toyed around with the parameters and the one that I've set to in the last commit seems to work best(at least in the sense of retrieving mu).
There are a few interesting observations:

1. Regardless of the number of chains, khat for each chain would all return infinity. This means that the result would be calculated from only the last W iterates, instead of the MCSE and ESS based cutoff.
2. There's quite a large difference in whether you run either 2 or 4 chains. For the std_normal model, running with 2 chains would terminate super fast(sets T_naught as 50 out of 1000 max iterations) and return the final mu as -0.4(terrible). On the other hand, 4 chains would run all the way up to 1000 iterations, and return -0.0009 as its final result. So I think running a minimum of 4 chains would be good for half-decent results.

bbbales2 commented:
Hmm, if it's running, on the test cases, my feeling is we should go ahead and pipe this out to the cmdstan level so we can start running full models on it (that'll give us more confidence that the basic algorithm is working, even as more stuff is added).

What possible flags do we need to pass in from the outside? I copy-pasted our current variational inference options below, I imagine we need to add more (like how many internal chains to use). Maybe we should remove some too. If you can write down what extra arguments we need (and default values and possible ranges), I can put up a cmdstan branch that has the stuff in it (and once you see the example you'll know how to change it yourself) and this way we can have a full working cmdstan:

```
variational
      Variational inference
      Valid subarguments: algorithm, iter, grad_samples, elbo_samples, eta, adapt, tol_rel_obj, eval_elbo, output_samples

      algorithm=<list element>
        Variational inference algorithm
        Valid values: meanfield, fullrank
        Defaults to meanfield

        meanfield
          mean-field approximation

        fullrank
          full-rank covariance

      iter=<int>
        Maximum number of ADVI iterations.
        Valid values: 0 < iter
        Defaults to 10000

      grad_samples=<int>
        Number of Monte Carlo draws for computing the gradient.
        Valid values: 0 < num_samples
        Defaults to 1

      elbo_samples=<int>
        Number of Monte Carlo draws for estimate of ELBO.
        Valid values: 0 < num_samples
        Defaults to 100

      eta=<double>
        Stepsize scaling parameter.
        Valid values: 0 < eta
        Defaults to 1

      adapt
        Eta Adaptation for Variational Inference
        Valid subarguments: engaged, iter

        engaged=<boolean>
          Boolean flag for eta adaptation.
          Valid values: [0, 1]
          Defaults to 1

        iter=<int>
          Number of iterations for eta adaptation.
          Valid values: 0 < iter
          Defaults to 50

      tol_rel_obj=<double>
        Relative tolerance parameter for convergence.
        Valid values: 0 <= tol
        Defaults to 0.01

      eval_elbo=<int>
        Number of iterations between ELBO evaluations
        Valid values: 0 < eval_elbo
        Defaults to 100

      output_samples=<int>
        Number of approximate posterior output draws to save.
        Valid values: 0 < output_samples
        Defaults to 1000
```

Dashadower commented:
Flags up to `adapt` would all be needed. Tol_rel_obj and eval_elbo would be replaced with the following:

- max_runs = 10000
    maximum number of iterations
- eval_window = 100
    Evaluation window length, and iterations between evaluations
- rhat_cut = 1.1
    Rhat threshold to stop preliminary optimization loop
- mcse_cut = 20
    Mcse threshold to set iterate averaging limit
- ess_cut = 20
    Ess threshold to set iterate averaging limit
- num_chains = 4
    Number of optimization chains to run
- adapt_iterations = 100
    Number of iterations to set initial stepsize

bbbales2 committed:
Merge remote-tracking branch 'origin/develop' into feature/rvi

bbbales2 committed:
Updating interface to pass arguments/work with robust vi

bbbales2 committed:
Updated rvi tests

Commit:
Update khat calculation

hyunjimoon commented on src/stan/variational/advi.hpp:
```diff
@@ -806,11 +806,20 @@ class advi {
+      for (int n = 0; n < n_tail; ++n) {
+        w_tail(n) = exp(lw_vec(lw_vec.size() - n_tail + n)) - exp(lw_vec(lw_vec.size() - n_tail));
+      }
+      stan::analyze::gpdfit(w_tail, khat, sigma);
```
added [this](https://github.com/stan-dev/loo/blob/51fb2d0511a765407ff1bd47fab7e79b5c62cbfb/R/psis.R#L209) logic from loo package, but getting minus khat values as follows:
```
Chain 0khat: 0.305205
Chain 1khat: -0.8578
Chain 2khat: 0.194817
Chain 3khat: -0.531074
```

Dashadower replied:
If you need the last ` n_tail` values in a vector,  you can use ` lw_vec.tail(n_tail)`

Dashadower committed:
Merge branch 'develop' of https://github.com/stan-dev/stan into feature/rvi

Dashadower committed:
Fix Secondary(post-convergence) iteration loop logic

Dashadower committed:
Manually update math sumodule

Dashadower committed:
Merge branch 'develop' of https://github.com/stan-dev/stan into feature/rvi

Dashadower committed:
Add split ESS and MCSE calculation to secondary iteration, clean up multiple stringstreams.

Dashadower committed:
Add window size to rvi and update relevant services

Dashadower committed:
Fix type

Dashadower committed:
Use split rhat for num_chains > 1

Dashadower committed:
Fix typo in split rhat

Dashadower commented on src/stan/variational/advi.hpp:
```diff
@@ -522,36 +381,369 @@ class advi {
+      double sigma, max_lw;
+      int n_tail;
+      if(n_posterior_samples_ < 225) {
+        n_tail = int(lw_vec.size() * 0.2);
```
@hyunjimoon 
I'm not sure how I should change this.
[loo](https://github.com/stan-dev/loo/blob/51fb2d0511a765407ff1bd47fab7e79b5c62cbfb/R/psis.R#L276) uses `min(0.2 * N, 3 * sqrt(N / ess))`

Dashadower committed:
Update rhat calculation

Dashadower committed:
Update math submodule.

bbbales2 committed:
Merge remote-tracking branch 'origin/develop' into feature/rvi

Commit:
Low-rank ADVI implementation

Implement low-rank ADVI. Initial pass at unit test cases and service
interface required for CmdStan implementation

Dashadower committed:
Fix Rhat calculation logic

bbbales2 committed:
Merge branch 'feature/rvi' of github.com:Dashadower/stan into feature/rvi

bbbales2 committed:
Merge remote-tracking branch 'origin/develop' into feature/rvi

Dashadower committed:
Remove run_rvi() and log per-chain parameter draws

Dashadower committed:
Merge branch 'feature/rvi' of https://github.com/dashadower/stan into feature/rvi

Dashadower committed:
Remove existing single chain(averaged) output from RVI.

Commit:
Lint & test fixes

yashikno committed:
Merge commit '5eed47e450ce44015228af9fcc0f9afee6d1e626' into HEAD

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

Commit:
Documentation

Commit:
Documentation

Commit:
Documentation and default

Commit:
Merge branch 'feature/issue-2750-advi-lowrank' of https://github.com/wjn0/stan into feature/issue-2750-advi-lowrank

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

Dashadower committed:
Change int to floating point type

Dashadower committed:
Merge branch 'develop' of https://github.com/stan-dev/stan into feature/rvi

stephensrmmartin committed:
Merge remote-tracking branch 'lowrank/feature/issue-2750-advi-lowrank' into feature/advi_lowrank

stephensrmmartin committed:
Merge remote-tracking branch 'rvi/feature/rvi' into feature/rvi

stephensrmmartin committed:
Merge branch 'feature/rvi' into feature/advi_lowrank

stephensrmmartin committed:
Change interface.

stephensrmmartin committed:
Change call to cmd_advi

bbbales2 committed:
Merge remote-tracking branch 'origin' into feature/rvi

stephensrmmartin committed:
Need to implement: set_approx_params, return_approx_params.

bbbales2 committed:
Cast int to double

stephensrmmartin committed:
Implemented virtual fns:

- set_approx_params
- return_approx_params
- num_approx_params

stephensrmmartin committed:
Typo: ind -> int

stephensrmmartin committed:
Fixed constructor name.

stephensrmmartin committed:
Change Q() to init_variational() in advi.

stephensrmmartin committed:
Changed Q() to init_variational()

stephensrmmartin committed:
Whitespace debugging.

stephensrmmartin committed:
int -> size_t

stephensrmmartin committed:
Removed int eval_elbo from advi(advi_base).

stephensrmmartin committed:
Merge branch 'feature/rvi' of https://github.com/Dashadower/stan into feature/rvi

stephensrmmartin committed:
Merge branch 'feature/rvi' into develop

stephensrmmartin committed:
Fix: lp_approx/log_g__ containing chain_ids

Added "chain_id__" name to lowrank service.

stephensrmmartin committed:
Version bump.

Dashadower committed:
Merge pull request #4 from stephensrmmartin/develop

Merge lowrank with feature/rvi

Dashadower committed:
Merge remote-tracking branch 'upstream/develop' into feature/rvi

Dashadower committed:
Update math submodule

Dashadower committed:
Add improvements to RVI.

Dashadower committed:
Update services

Dashadower committed:
Change to lowrak services

Dashadower committed:
Remove debug code from advi

Dashadower committed:
Remove debug code

Dashadower committed:
Remove trailing newline when writing to output csv file.

Dashadower committed:
Merge branch 'develop' of https://github.com/stan-dev/stan into feature/rvi
